python/pyspark/ml/tuning.py

```diff
@@ -207,6 +210,205 @@ def _to_java_impl(self):
         return java_estimator, java_epms, java_evaluator
 
 
+class _ValidatorSharedReadWrite:
+
+    @staticmethod
+    def saveImpl(path, instance, sc, extraMetadata=None):
+        from pyspark.ml.classification import OneVsRest
+        numParamsNotJson = 0
+        jsonEstimatorParamMaps = []
+        for paramMap in instance.getEstimatorParamMaps():
+            jsonParamMap = []
+            for p, v in paramMap.items():
+                jsonParam = {'parent': p.parent, 'name': p.name}
+                if (isinstance(v, Estimator) and not (
+                        isinstance(v, _ValidatorParams) or
+                        isinstance(v, OneVsRest))
+                    ) or isinstance(v, Transformer) or \
+                        isinstance(Evaluator):
+                    relative_path = f'epm_{p.name}{numParamsNotJson}'
+                    param_path = os.path.join(path, relative_path)
+                    numParamsNotJson += 1
+                    v.save(param_path)
+                    jsonParam['value'] = relative_path
+                    jsonParam['isJson'] = False
+                elif isinstance(v, MLWritable):
+                    raise RuntimeError(
+                        "ValidatorSharedReadWrite.saveImpl does not handle parameters of type: "
+                        "MLWritable that are not Estimaor/Evaluator/Transformer, and if parameter is estimator,"
+                        "it cannot be Validator or OneVsRest")
```

Comment:
In JVM side, the logic is only support `DefaultParamsWritable` instance. 
Pyspark side we cannot use it, but we can equivalently use:
Support Estimaor/Evaluator/Transformer type but excludes Validator or OneVsRest. These types in JVM side are inherits `DefaultParamsWritable`.

Reply:
It would be clearer if we can create an interface similar to `DefaultParamsWritable` and do not use
~~~
               if (isinstance(v, Estimator) and not (
                        isinstance(v, _ValidatorParams) or
                        isinstance(v, OneVsRest))
                    ) or isinstance(v, Transformer) or \
                        isinstance(Evaluator):
~~~
since it looks very confusing.

Reply:
Why support `Evaluator`? It seems that it's estimator param maps, not evaluator param maps.
Why don't support `OneVsRest`? Is it not supported for tuning or just to keep consistent with the scala side? We may just save the `OneVsRest.classifier` the same way as we save the estimator.

Reply:
Keep consistent with java side. Java side support `Evaluator` type param. Although current spark estimators don't have such params, but 3rd party estimator may has that.